File: two-pointers/med-0015-3sum/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); // Sort the array
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            // Skip duplicates
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int a = nums[i];
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = a + nums[left] + nums[right];

                if (sum < 0) {
                    left++;
                } else if (sum > 0) {
                    right--;
                } else {
                    // Found a triplet
                    result.add(Arrays.asList(a, nums[left], nums[right]));

                    int leftVal = nums[left], rightVal = nums[right];
                    // Skip duplicates
                    while (left < right && nums[left] == leftVal) left++;
                    while (left < right && nums[right] == rightVal) right--;
                }
            }
        }

        return result;
    }
}

File: two-pointers/med-0015-3sum/Solution.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            a = nums[i]
            l, r = i + 1, len(nums) - 1

            while l < r:
                curr_sum = a + nums[l] + nums[r]
                if curr_sum < 0:
                    l += 1
                elif curr_sum > 0:
                    r -= 1
                else:
                    result.append([a, nums[l], nums[r]])
                    l_val, r_val = nums[l], nums[r]
                    while l < r and nums[l] == l_val:
                        l += 1
                    while l < r and nums[r] == r_val:
                        r -= 1

        return result
